include/exadg/incompressible_navier_stokes/spatial_discretization/generalized_newtonian_model.cpp

```diff
@@ -0,0 +1,334 @@
+      tensor velocity_gradient          = integrator_m.get_gradient(q);
+      tensor velocity_gradient_neighbor = integrator_p.get_gradient(q);
+
+      // get the coefficients
+      scalar viscosity          = Base::viscous_kernel->get_coefficient_face(face, q);
+      scalar viscosity_neighbor = Base::viscous_kernel->get_coefficient_face_neighbor(face, q);
+
+      add_generalized_newtonian_viscosity(viscosity, velocity_gradient);
+      add_generalized_newtonian_viscosity(viscosity_neighbor, velocity_gradient_neighbor);
+
+      // set the coefficients
+      Base::viscous_kernel->set_coefficient_face(face, q, viscosity);
+      Base::viscous_kernel->set_coefficient_face_neighbor(face, q, viscosity_neighbor);
+    }
+  }
+}
+
+template<int dim, typename Number>
+void
+GeneralizedNewtonianModel<dim, Number>::boundary_face_loop_set_coefficients(
+  dealii::MatrixFree<dim, Number> const & matrix_free,
+  VectorType &,
+  VectorType const & src,
+  Range const &      face_range) const
+{
+  FaceIntegratorU integrator(matrix_free,
+                             true,
+                             Base::dof_index_velocity,
+                             Base::quad_index_velocity_linear);
+
+  // loop over all boundary faces
+  for(unsigned int face = face_range.first; face < face_range.second; face++)
+  {
+    integrator.reinit(face);
+    integrator.read_dof_values(src);
+
+    // we only need the gradient
+    integrator.evaluate(dealii::EvaluationFlags::gradients);
+
+    // loop over all quadrature points
+    for(unsigned int q = 0; q < integrator.n_q_points; ++q)
+    {
+      // calculate velocity gradient
+      tensor velocity_gradient = integrator.get_gradient(q);
+
+      // get the coefficients
+      scalar viscosity = Base::viscous_kernel->get_coefficient_face(face, q);
+
+      add_generalized_newtonian_viscosity(viscosity, velocity_gradient);
+
+      // set the coefficients
+      Base::viscous_kernel->set_coefficient_face(face, q, viscosity);
+    }
+  }
+}
+
+template<int dim, typename Number>
+void
+GeneralizedNewtonianModel<dim, Number>::add_generalized_newtonian_viscosity(
+  scalar &       viscosity,
+  tensor const & velocity_gradient) const
+{
+  tensor symmetric_velocity_gradient = 0.5 * (velocity_gradient + transpose(velocity_gradient));
+
+  scalar shear_rate =
+    std::sqrt(2.0 * scalar_product(symmetric_velocity_gradient, symmetric_velocity_gradient));
+
+  scalar viscosity_factor;
+  compute_viscosity_factor(viscosity_factor, shear_rate);
+
+  viscosity += viscosity_factor * viscosity_margin;
+}
+
+template<int dim, typename Number>
+void
+GeneralizedNewtonianModel<dim, Number>::compute_viscosity_factor(scalar &       viscosity_factor,
+                                                                 scalar const & shear_rate) const
+{
+  switch(generalized_newtonian_model)
+  {
+    case GeneralizedNewtonianViscosityModel::Undefined:
```

Comment:
I would implement the general formula as a lambda function here (potentially with parameters for k, n, a). The switch-cases below then call this lambda function, or implement the function directly if it is much simpler than the general function (because of special parameter choices).